src/Microsoft.FeatureManagement/Targeting/ContextualTargetingFeatureVariantAssigner.cs

```diff
@@ -0,0 +1,143 @@
+        /// <param name="targetingContext">The targeting context used to determine which variant should be assigned.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns></returns>
+        public ValueTask<FeatureVariant> AssignVariantAsync(FeatureVariantAssignmentContext variantAssignmentContext, ITargetingContext targetingContext, CancellationToken cancellationToken)
+        {
+            if (variantAssignmentContext == null)
+            {
+                throw new ArgumentNullException(nameof(variantAssignmentContext));
+            }
+
+            if (targetingContext == null)
+            {
+                throw new ArgumentNullException(nameof(targetingContext));
+            }
+
+            FeatureDefinition featureDefinition = variantAssignmentContext.FeatureDefinition;
+
+            if (featureDefinition == null)
+            {
+                throw new ArgumentNullException(nameof(variantAssignmentContext.FeatureDefinition));
+            }
+
+            if (featureDefinition.Variants == null)
+            {
+                throw new ArgumentNullException(nameof(featureDefinition.Variants));
+            }
+
+            FeatureVariant variant = null;
+
+            FeatureVariant defaultVariant = null;
+
+            double cumulativePercentage = 0;
+
+            var cumulativeGroups = new Dictionary<string, double>();
+
+            foreach (FeatureVariant v in featureDefinition.Variants)
+            {
+                if (defaultVariant == null && v.Default)
+                {
+                    defaultVariant = v;
+                }
+
+                TargetingFilterSettings targetingSettings = v.AssignmentParameters.Get<TargetingFilterSettings>();
+
+                if (targetingSettings == null)
+                {
+                    if (v.Default)
+                    {
+                        //
+                        // Valid to omit audience for default variant
+                        continue;
+                    }
+                    else
+                    {
+                        targetingSettings = new TargetingFilterSettings();
+                    }
+                }
+
+                if (!TargetingEvaluator.TryValidateSettings(targetingSettings, out string paramName, out string reason))
+                {
+                    throw new ArgumentException(reason, paramName);
+                }
+
+                AccumulateAudience(targetingSettings.Audience, cumulativeGroups, ref cumulativePercentage);
+
+                if (TargetingEvaluator.IsTargeted(targetingSettings, targetingContext, _options.IgnoreCase, featureDefinition.Name))
+                {
+                    variant = v;
+
+                    break;
+                }
+            }
+
+            return new ValueTask<FeatureVariant>(variant);
```

Comment:
Okay, so this is about the split of responsibilities. Since this is part of the extensibility model, we probably should document it somewhere if we decide this is the contract we want.

An assigner can handle the default case, but if it returns null, the default variant will be returned to the user. We choose not to handle the default variant in this assigner (but nothing wrong to do it).

I know our ConfigurationFeatureVariantOptionsResolver requires the variant is not null. If I have a custom IFeatureVariantOptionsResolver and a custom assigner that relies on the resolver to handle the null variant, I would not be able to do it with current design. Maybe I'm over thinking. Just think loudly to get your thoughts :)

Reply:
I consider default variant fallback a feature of the system. I think the appropriate way to handle the scenario you described is to configure the dynamic features being evaluated with your custom assigner to have a default variant that is handled as desired by the custom IFeatureVariantOptionsResolver. Rather than resolving a null variant, it is resolving a variant that is deliberately provided.

Reply:
Yes, that's more deterministic. Sounds like a plan.

Can we mention somewhere in the comments that the default variant will be returned if an assigner returns null?